sdk/eventgrid/Azure.Messaging.EventGrid/src/Customization/EventGridConsumer.cs

```diff
@@ -0,0 +1,300 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Text;
+using System.Text.Json;
+using System.Threading;
+using System.Threading.Tasks;
+using Azure.Core;
+using Azure.Core.Pipeline;
+using Azure.Messaging.EventGrid.Models;
+
+namespace Azure.Messaging.EventGrid
+{
+    /// <summary>
+    /// Class used to decode events from the Event Grid service.
+    /// </summary>
+    public class EventGridConsumer
+    {
+        private readonly ObjectSerializer _objectSerializer;
+        private readonly IDictionary<string, Type> _customEventTypeMappings;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="EventGridConsumer"/> class.
+        /// </summary>
+        public EventGridConsumer() :
+            this(new EventGridConsumerOptions())
+        {
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="EventGridConsumer"/> class.
+        /// </summary>
+        /// <param name="options"> Options for configuring deserialization. </param>
+        public EventGridConsumer(EventGridConsumerOptions options)
+        {
+            Argument.AssertNotNull(options, nameof(options));
+            _objectSerializer = options.DataSerializer;
+            _customEventTypeMappings = options.CustomEventTypeMappings;
```

Comment:
since this is a mutable ref type, maybe we should create a new Dictionary based on the options to keep the client immutable.